File: java/amp-commanding/src/main/java/amp/commanding/CommandRegistration.java
package amp.commanding;


import java.util.HashMap;
import java.util.List;
import java.util.Map;

import cmf.bus.Envelope;
import cmf.bus.EnvelopeHeaderConstants;
import cmf.bus.IEnvelopeFilterPredicate;
import cmf.bus.IRegistration;
import com.google.common.base.Function;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


/**
 * Created with IntelliJ IDEA.
 * User: jar349
 * Date: 5/1/13
 */
public class CommandRegistration implements IRegistration {

    private static final Logger LOG = LoggerFactory.getLogger(CommandRegistration.class);

    private ICommandChainProcessor _processor;
    private List<ICommandProcessor> _processingChain;
    private ICommandHandler _handler;
    private Map<String, String> _regInfo;


    @Override
    public Map<String, String> getRegistrationInfo() { return _regInfo; }

    @Override
    public IEnvelopeFilterPredicate getFilterPredicate() { return null; }


    public CommandRegistration(
            ICommandChainProcessor processor,
            List<ICommandProcessor> processingChain,
            ICommandHandler handler) {

        _processor = processor;
        _processingChain = processingChain;
        _handler = handler;

        _regInfo = new HashMap<String, String>();
        _regInfo.put(EnvelopeHeaderConstants.MESSAGE_TOPIC, _handler.getCommandType().getCanonicalName());
        _regInfo.put(EnvelopeHeaderConstants.MESSAGE_TYPE, _handler.getCommandType().getCanonicalName());
    }


    @Override
    public Object handle(final Envelope env) throws Exception {

        try {
            // create a context to send through the processors
            final CommandContext ctx = new CommandContext(CommandContext.Directions.In, env);

            _processor.processCommand(ctx, _processingChain, new IContinuationCallback() {

                @Override
                public void continueProcessing() throws CommandException {
                    _handler.handle(ctx.getCommand(), env.getHeaders());
                }
            });

        }
        catch (CommandException ex) {
            String message = "Failed to process an incoming command envelope.";
            LOG.error(message, ex);
            throw new Exception(message, ex);
        }

        return null;
    }

    @Override
    public Object handleFailed(Envelope env, Exception ex) throws Exception {
        return null;
    }
}

File: dotnet/source/amp.commanding/CommandRegistration.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Common.Logging;
using cmf.bus;

namespace amp.commanding
{
    public class CommandRegistration : IRegistration
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof(CommandRegistration));

        private readonly ICommandHandler _handler;
        private readonly Predicate<Envelope> _filterPredicate;
        private readonly Func<Envelope, CommandContext> _envelopeOpener;


        public Predicate<Envelope> Filter
        {
            get { return _filterPredicate; }
        }

        public object Handle(Envelope env)
        {
            Log.Debug("Enter CommandRegistration # handle( Envelope env )");

            try {
                // use the provided envelope opener to get a context
                CommandContext context = _envelopeOpener(env);

                if (null != context)
                {
                    _handler.Handle(context.Command, context.Envelope.Headers);
                }
                else
                {
                    // if our opener returned null, that signifies that it declined
                    // to open it, and that we should quietly drop this envelope
                    // So, I'll return null here to make this explicit, even though
                    // just doing nothing here would lead to the 'return null' that
                    // is just a few lines below this.
                    return null;
                }
            }
            catch (CommandException ex) {
                String message = "Failed to process an incoming command envelope.";
                Log.Error(message, ex);
                throw new Exception(message, ex);
            }

            return null;
        }

        public object HandleFailed(Envelope env, Exception ex)
        {
            Log.Error("Failed to handle a command.", ex);
            return null;
        }

        public IDictionary<string, string> Info { get; set; }



        public CommandRegistration(Func<Envelope, CommandContext> envelopeOpener, ICommandHandler handler)
        {
            _envelopeOpener = envelopeOpener;
            _handler = handler;

            string handledType = _handler.CommandType.FullName;

            this.Info = new Dictionary<string, string>
            {
                {EnvelopeHeaderConstants.MESSAGE_TOPIC, handledType},
                {EnvelopeHeaderConstants.MESSAGE_TYPE, handledType}
            };
        }

        public CommandRegistration(
            Func<Envelope, CommandContext> envelopeOpener,
            ICommandHandler handler,
            Predicate<Envelope> filterPredicate)
            : this(envelopeOpener, handler)
        {
            _filterPredicate = filterPredicate;
        }
    }
}
